File: lib/join/oraclejoin/rules.go
// Teleport
// Copyright (C) 2025 Gravitational, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package oraclejoin

import (
	"slices"

	"github.com/gravitational/trace"

	"github.com/gravitational/teleport/api/types"
	"github.com/gravitational/teleport/lib/join/provision"
)

// CheckOracleAllowRules checks that a the oracle rules in a provision token
// allow an OCI instance with the given claims to join the cluster.
func CheckOracleAllowRules(claims *Claims, token provision.Token) error {
	instanceRegion := claims.Region()
	for _, rule := range token.GetOracle().Allow {
		if rule.Tenancy != claims.TenancyID {
			// rule.Tenancy is required, if it doesn't match the instance skip this rule.
			continue
		}
		if !ruleAllowsCompartment(rule, claims.CompartmentID) {
			// Skip this rule if it does not allow the instance's compartment.
			continue
		}
		if !ruleAllowsRegion(rule, instanceRegion) {
			// Skip this rule if it does not allow the instance's region.
			continue
		}
		if !ruleAllowsInstanceID(rule, claims.InstanceID) {
			// Skip this rule if it does not allow the instance's ID.
			continue
		}
		// This rule allows the instance to join.
		return nil
	}
	return trace.AccessDenied("instance %v did not match any allow rules in token %v", claims.InstanceID, token.GetName())
}

func ruleAllowsCompartment(rule *types.ProvisionTokenSpecV2Oracle_Rule, instanceCompartment string) bool {
	if len(rule.ParentCompartments) == 0 {
		// Empty list means all compartments are allowed.
		return true
	}
	return slices.Contains(rule.ParentCompartments, instanceCompartment)
}

func ruleAllowsRegion(rule *types.ProvisionTokenSpecV2Oracle_Rule, instanceRegion string) bool {
	if len(rule.Regions) == 0 {
		// Empty list means all regions are allowed.
		return true
	}
	return slices.ContainsFunc(rule.Regions, func(allowedRegion string) bool {
		canonicalAllowedRegion, _ := ParseRegion(allowedRegion)
		return canonicalAllowedRegion == instanceRegion
	})
}

func ruleAllowsInstanceID(rule *types.ProvisionTokenSpecV2Oracle_Rule, instanceID string) bool {
	if len(rule.Instances) == 0 {
		// Empty list means all instance IDs are allowed.
		return true
	}
	return slices.Contains(rule.Instances, instanceID)
}

File: web/packages/shared/components/ScheduleEditor/rules.ts
/**
 * Teleport
 * Copyright (C) 2025  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Rule } from 'shared/components/Validation/rules';

import { timeOptionsAll } from './const';
import { Schedule, Shift } from './types';

/**
 * validSchedule validates a schedule.
 * A schedule is expected to contain at least one shift, and all shifts
 * must be valid.
 *
 * @param schedule - The target schedule to validate.
 * @returns a validator function that ensure the schedule is valid.
 */
export const validSchedule: Rule<Schedule> = (schedule: Schedule) => () => {
  const enabledShifts = Object.values(schedule.shifts).filter(shift => !!shift);

  if (enabledShifts.length === 0) {
    return {
      valid: false,
      message: `At least one shift is required.`,
    };
  }

  for (const shift of enabledShifts) {
    const error = validateShift(shift);
    if (error) {
      return {
        valid: false,
        message: `Shift must be between 12:00AM and 11:59PM.`,
      };
    }
  }

  return {
    valid: true,
  };
};

/**
 * validShift validates a shift.
 * A shift is expected to be between 00:00 and 23:59.
 *
 * @param shift - The target shift to validate.
 * @returns a validator function that ensure the shift is valid.
 */
export const validShift: Rule<Shift> = (shift: Shift) => () => {
  const error = validateShift(shift);
  return {
    valid: !error,
    message: error,
  };
};

const validateShift = (shift: Shift) => {
  const start = shift.startTime?.value;
  const end = shift.endTime?.value;

  if (!start || !end) {
    return 'start and end interval required';
  }

  const startIndex = timeOptionsAll.findIndex(t => t.value === start);
  const endIndex = timeOptionsAll.findIndex(t => t.value === end);

  if (startIndex === -1 || endIndex === -1) {
    return 'invalid time';
  }

  if (startIndex >= endIndex) {
    return 'start time must be before end time';
  }
  return undefined;
};
